bin/elastishadow/main.c

```diff
@@ -0,0 +1,1261 @@
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/command_line_parser.h>
+#include <aws/common/condition_variable.h>
+#include <aws/common/hash_table.h>
+#include <aws/common/log_channel.h>
+#include <aws/common/mutex.h>
+#include <aws/common/string.h>
+#include <aws/common/uuid.h>
+
+#include <aws/io/channel_bootstrap.h>
+#include <aws/io/event_loop.h>
+#include <aws/io/socket.h>
+#include <aws/io/tls_channel_handler.h>
+#include <aws/io/uri.h>
+
+#include <aws/mqtt/private/shared.h>
+#include <aws/mqtt/request-response/request_response_client.h>
+#include <aws/mqtt/v5/mqtt5_client.h>
+#include <aws/mqtt/v5/mqtt5_types.h>
+
+#include <inttypes.h>
+
+#ifdef _MSC_VER
+#    pragma warning(disable : 4996) /* Disable warnings about fopen() being insecure */
+#    pragma warning(disable : 4204) /* Declared initializers */
+#    pragma warning(disable : 4221) /* Local var in declared initializer */
+#endif
+
+#ifdef WIN32
+// Windows does not need specific imports
+#else
+#    include <stdio.h>
+#endif
+
+struct app_ctx {
+    struct aws_allocator *allocator;
+    struct aws_mutex lock;
+    struct aws_condition_variable signal;
+    struct aws_uri uri;
+    uint32_t port;
+    const char *cert;
+    const char *key;
+
+    struct aws_tls_connection_options tls_connection_options;
+
+    const char *log_filename;
+    enum aws_log_level log_level;
+
+    struct aws_mqtt5_client *client;
+    struct aws_mqtt_request_response_client *rr_client;
+
+    /*
+     * &aws_shadow_streaming_operation.id -> aws_shadow_streaming_operation *
+     */
+    struct aws_hash_table streaming_operations;
+    uint64_t next_id;
+};
+
+struct aws_shadow_streaming_operation {
+    struct aws_allocator *allocator;
+
+    struct aws_byte_buf thing;
+    struct aws_byte_buf shadow;
+    struct aws_byte_buf topic_filter;
+
+    uint64_t id;
+    struct aws_mqtt_rr_client_operation *streaming_operation;
+};
+
+struct aws_shadow_streaming_operation *s_aws_shadow_streaming_operation_new(
+    struct aws_allocator *allocator,
+    struct app_ctx *app_ctx,
+    struct aws_byte_cursor thing,
+    struct aws_byte_cursor shadow,
+    struct aws_byte_cursor topic_filter) {
+    struct aws_shadow_streaming_operation *operation =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_shadow_streaming_operation));
+
+    operation->allocator = allocator;
+    operation->id = ++app_ctx->next_id;
+    operation->streaming_operation = NULL;
+
+    aws_byte_buf_init_copy_from_cursor(&operation->thing, allocator, thing);
+    aws_byte_buf_init_copy_from_cursor(&operation->shadow, allocator, shadow);
+    aws_byte_buf_init_copy_from_cursor(&operation->topic_filter, allocator, topic_filter);
+
+    return operation;
+}
+
+void s_aws_shadow_streaming_operation_destroy(struct aws_shadow_streaming_operation *operation) {
+    if (operation == NULL) {
+        return;
+    }
+
+    aws_byte_buf_clean_up(&operation->thing);
+    aws_byte_buf_clean_up(&operation->shadow);
+
```

Comment:
May be intentional or handled elsewhere later but noting now that we're not cleaning up the topic_filter byte buf in this destroy.

Reply:
Yes, that was missed.